server/mailrepository/mailrepository-memory/src/main/java/org/apache/james/mailrepository/memory/MemoryMailRepositoryPropertiesStore.java

```diff
@@ -0,0 +1,51 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailrepository.memory;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.james.mailrepository.api.MailRepositoryProperties;
+import org.apache.james.mailrepository.api.MailRepositoryPropertiesStore;
+import org.apache.james.mailrepository.api.MailRepositoryUrl;
+import org.reactivestreams.Publisher;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import reactor.core.publisher.Mono;
+
+public class MemoryMailRepositoryPropertiesStore implements MailRepositoryPropertiesStore {
+
+    private final ConcurrentHashMap<MailRepositoryUrl, MailRepositoryProperties> propertiesMap;
+
+    @VisibleForTesting
+    MemoryMailRepositoryPropertiesStore() {
+        this.propertiesMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public Publisher<Void> store(MailRepositoryUrl url, MailRepositoryProperties properties) {
+        return Mono.fromRunnable(() -> propertiesMap.put(url, properties));
+    }
+
+    @Override
+    public Publisher<MailRepositoryProperties> retrieve(MailRepositoryUrl url) {
+        return Mono.fromSupplier(() -> propertiesMap.get(url));
```

Comment:
Should return a Publisher<Optional<MailRepositoryProperties>>

Reply:
the implementation of Publisher<T> here is the Mono<T>. And it does care about the empty-ness. The callers are able to retrieve Optional<T> instances:
```java
Optional<MailRepositoryProperties> properites = Mono.from(store.retrieve(url))
      .blockOptional();
```

Reply:
And if the get on the map return a null value, it's a problem because a null value in a reactive stream will stop the stream processing and throw a NullPointerException as defined in the reactive-streams laws. 

https://github.com/reactive-streams/reactive-streams-jvm#user-content-1.9

Reply:
I think NPE happens only when the function in the `map(Fuction<T, R>)` returns a null value. In case of null result. The map operators after will never be reached because the current state of the Mono is empty. I wrote a test to prove there is no NPE even supplier `() -> propertiesMap.get(url)` return a null value. please have a look at 818d9bc

Reply:
Ok

Reply:
let's wait for other member opinions, I still keep the idea that Mono already cares about empty-ness.

Reply:
It's our 'policy' to manage emptiness-ability as part as `Mono`.

I'm more concerned about throwing an NPE in your `Publisher`.

I'd prefer something like:

```
Mono.defer(() -> Mono.justOrEmpty(propertiesMap.get(url)))
    .switchIfEmpty(Mono.error("Can not ..."));
```

Reply:
`.switchIfEmpty(Mono.error("Can not ..."))` will make client feel awkward when then don't want an error mono in case of not thing found. Again, I think NPE throwing will never happen with my implementation. Mono.fromSupplier() care about the empty-ness of the supplier. Unless I the clients are irresponsible to add more function to the pipeline. E.g.
```java
Mono.from(store.retrieve(url)).block(); // will never throw
```
```java
Mono.from(store.retrieve(url))
    .map(result1 -> null)
    .map(result2 -> result2....)  // this will throw but since the first map operation, it's is the responsibility of the clients to maintain the pipeline.
    .block();
```

As your implementation, it could be very well produce a NPE with un-carefull code implementation like above.
```java
Publisher retrieve(MailRepositoryUrl url) {
    return Mono.defer(() -> Mono.justOrEmpty(propertiesMap.get(url)))
        .switchIfEmpty(Mono.error("Can not ..."));
}

Mono.from(store.retrieve(url)) // assuming the mono is not empty 
     .map(result1 -> null)
    .map(result2 -> result2....)  // still throw
    .block();
```

Reply:
Ok then.